### PR TITLE
Symbol grammar clarifications

### PR DESCRIPTION
**Pull request by kritzcreek**
- Forbids empty or 'local' scheme (this removes ambiguities from the grammar)
- Unifies "any UTF-8" wording
- Specifies that identifiers that _can_ be encoded as simple identifiers, have to be encoded as simple identifiers

Unfortunately I can not get the "build process" to work on my local machine.

### Test plan

Added checks to the `lint` command for checking canonical identifiers and that parsing succeeds

### CODE REVIEW

**Review by varungandhi-src — APPROVED** [review]

**Comment by varungandhi-src** [comment]
> We could add checks to `scip lint`

Sure, that'd be great.

**Comment by kritzcreek** [comment]
> Sure, that'd be great.

Allright, I think that should work? PTAL

File: bindings/go/scip/symbol.go [hunk 1]
```diff
@@ -22,10 +21,13 @@ func IsLocalSymbol(symbol string) bool {
 	return strings.HasPrefix(symbol, "local ")
 }
 
+func isSimpleIdentifierCharacter(c rune) bool {
+	return ('0' <= c && c <= '9') || ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || c == '$' || c == '+' || c == '-' || c == '_'
```
Comment (varungandhi-src) [inline bindings/go/scip/symbol.go:25]:
```suggestion
	return c == '_' || c == '+' || c == '-' || c == '$' || ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || ('0' <= c && c <= '9')
```

nit: Re-arranging to match up with order in doc comment in spec

File: cmd/scip/lint.go [hunk 2]
```diff
@@ -236,6 +240,22 @@ func (st *symbolTable) addOccurrence(path string, occ *scip.Occurrence) error {
 	return nil
 }
 
+func lintSymbolString(symbol string, context string) error {
+	if symbol == "" {
+		return emptyStringError{what: "symbol", context: context}
+	}
+	sym, err := scip.ParseSymbol(symbol)
+	if err != nil {
+		// TODO: This should be linted (but it makes all the tests fail)
```
Comment (varungandhi-src) [inline cmd/scip/lint.go:249]: Writing the long symbol names makes writing tests tedious, ideally we'd have a config object to disable some checks in tests. Maybe you can use a global variable here? It'd be off by default, and only one test would set it when checking for this code path. The other tests wouldn't need to worry about producing grammatically correct symbol names.

**Review by varungandhi-src — APPROVED** [review]
Added some minor non-blocking comments

File: cmd/scip/lint.go (hunk 2 above)
Comment (kritzcreek) [reply cmd/scip/lint.go:249]: What about introducing a constructor function for the tests? Something like `sym("a")` that expands to `. . . . a#`?

**Comment by kritzcreek** [comment]
Noticed one more difference between the spec and `scip-clang` while writing some benchmarks. `scip-clang` will generate descriptors like `my_func().`, while the spec does not make the disambiguator in the parenthesis optional.

Spec issue, or scip-clang issue?

File: cmd/scip/lint.go (hunk 2 above)
Comment (varungandhi-src) [reply cmd/scip/lint.go:249]: Maybe we can move that logic to `makeIndex` or `parseSymbolInfo`? In hindsight, I'm not sure if adding a DSL to specify relationships was a good idea, but I didn't want to have large chains of functions/methods when writing test cases. 🙈

**Comment by varungandhi-src** [comment]
> Noticed one more difference between the spec and scip-clang while writing some benchmarks. scip-clang will generate descriptors like my_func()., while the spec does not make the disambiguator in the parenthesis optional.

Hmm, I don't see why the disambiguator should be forced to be non-empty -- the rationale for having the disambiguator is that different disambiguator values can correspond to different method overloads. An empty string is just another string, which is different from other non-empty strings. Unless someone has a compelling reason otherwise, I think we should allow the disambiguator to be empty.

File: cmd/scip/lint.go (hunk 2 above)
Comment (kritzcreek) [reply cmd/scip/lint.go]: Okay I tried... but the DSL was a bit too much too handle so I went with the global as you suggested :D

**Comment by kritzcreek** [comment]
> Unless someone has a compelling reason otherwise, I think we should allow the disambiguator to be empty.

@olafurpg @keynmol Thoughts?

**Comment by olafurpg** [comment]
+1 to allowing empty disambiguators

**Pull request merged**